File: packages/bridge/tests/websocket/connection_purpose.test.ts
/**
 * Tests for connection purpose metadata handling (probe vs primary).
 *
 * These tests define the expected behavior for distinguishing short-lived
 * probe connections from primary mobile sessions. Currently FAILING -
 * implementation code changes required.
 *
 * Connection Purpose Semantics (per bridge-protocol.md):
 * - Probe: ~500ms-2s duration, health verification, capability check, close code 1000
 * - Primary: Session lifetime, active session for all communication
 */
import { eventBus } from "../../src/notifications/event_bus";
import { ConnectionManager } from "../../src/websocket/connection_manager";
import { MessageHandler } from "../../src/websocket/message_handler";
import type { AgentSession, BridgeMessage, ConnectionAckPayload } from "../../src/types";

describe("Connection Purpose Metadata Handling", () => {
  afterEach(() => {
    eventBus.removeAllListeners();
  });

  /**
   * Helper to create test dependencies with configurable client state.
   */
  function createDependencies(
    options: {
      connectionMode?: "secure_remote" | "direct_public";
      existingPurpose?: "primary" | "probe";
    } = {},
  ) {
    const { connectionMode = "secure_remote", existingPurpose } = options;
    const sentMessages: Array<{ clientId: string; message: BridgeMessage }> = [];
    const connectionManager = {
      sendToClient: jest.fn((clientId: string, message: BridgeMessage) => {
        sentMessages.push({ clientId, message });
      }),
      authenticateClient: jest.fn(),
      acknowledgeWarning: jest.fn(),
      getClient: jest.fn(() => ({
        authenticated: true,
        sessionIds: [],
        connectionMode,
        connectionModeDescription: "Secure tunnel connection (test.trycloudflare.com)",
        bridgeUrl: "wss://test.trycloudflare.com",
        warningAcknowledged: false,
        purpose: existingPurpose,
      })),
      addSessionToClient: jest.fn(),
      getClientsForSession: jest.fn(),
      broadcast: jest.fn(),
      setClientPurpose: jest.fn(),
      getPrimaryClients: jest.fn(() => []),
      getProbeClients: jest.fn(() => []),
    };

    const agentSdkAdapter = {
      handleSessionStart: jest.fn(),
      handleMessage: jest.fn(),
      handleApprovalResponse: jest.fn(),
      handleSessionEnd: jest.fn(),
    };

    const activeSession: AgentSession = {
      id: "sess-1",
      agent: "claude-code",
      title: "project",
      model: "claude-opus-4-6",
      working_directory: "/repo/project",
      created_at: new Date().toISOString(),
      status: "idle",
    };

    const agentSessionManager = {
      getActiveSessions: jest.fn(() => [activeSession]),
      getSession: jest.fn((sessionId: string) =>
        sessionId === activeSession.id ? activeSession : undefined,
      ),
    };

    const gitService = {
      getStatus: jest.fn(),
      commit: jest.fn(),
      getDiff: jest.fn(),
    };

    const eventQueue = {
      replay: jest.fn(() => []),
      acknowledgeNotifications: jest.fn(),
    };

    const handler = new MessageHandler(
      connectionManager as never,
      agentSdkAdapter as never,
      agentSessionManager as never,
      gitService as never,
      eventQueue as never,
    );

    return {
      handler,
      sentMessages,
      connectionManager,
      agentSdkAdapter,
      activeSession,
      eventQueue,
    };
  }

  describe("Auth payload purpose field", () => {
    it("should accept purpose='probe' in auth payload during connection handshake", async () => {
      const { handler, sentMessages, connectionManager } = createDependencies();

      // Client sends auth with purpose: "probe" to check bridge health
      await handler.handle(
        "client-probe-1",
        JSON.stringify({
          type: "auth",
          id: "auth-probe-1",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            client_version: "1.0.0",
            platform: "ios",
            purpose: "probe", // NEW: Client declares this is a probe connection
          },
        }),
      );

      expect(connectionManager.authenticateClient).toHaveBeenCalledWith("client-probe-1");

      const ackMessage = sentMessages[0]?.message as BridgeMessage<ConnectionAckPayload>;
      expect(ackMessage.type).toBe("connection_ack");
      // NEW: Bridge should echo back accepted purpose
      expect(ackMessage.payload.purpose).toBe("probe");
    });

    it("should accept purpose='primary' in auth payload for main session connection", async () => {
      const { handler, sentMessages } = createDependencies();

      await handler.handle(
        "client-primary-1",
        JSON.stringify({
          type: "auth",
          id: "auth-primary-1",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            client_version: "1.0.0",
            platform: "android",
            purpose: "primary", // Primary connection for session
          },
        }),
      );

      const ackMessage = sentMessages[0]?.message as BridgeMessage<ConnectionAckPayload>;
      expect(ackMessage.type).toBe("connection_ack");
      expect(ackMessage.payload.purpose).toBe("primary");
    });

    it("should default to 'primary' purpose when purpose field is omitted", async () => {
      const { handler, sentMessages } = createDependencies();

      // Legacy clients without purpose field should default to primary
      await handler.handle(
        "client-legacy-1",
        JSON.stringify({
          type: "auth",
          id: "auth-legacy-1",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            client_version: "1.0.0",
            platform: "ios",
            // purpose field omitted - should default to 'primary'
          },
        }),
      );

      const ackMessage = sentMessages[0]?.message as BridgeMessage<ConnectionAckPayload>;
      expect(ackMessage.payload.purpose).toBe("primary");
    });
  });

  describe("ConnectionManager purpose tracking", () => {
    function addClient(connectionManager: ConnectionManager, clientId: string): void {
      connectionManager.addClient(clientId, {
        readyState: 1,
        send: jest.fn(),
      } as never);
    }

    it("should expose getPrimaryClients() returning only primary purpose clients", async () => {
      const connectionManager = new ConnectionManager();

      addClient(connectionManager, "client-1");
      addClient(connectionManager, "client-2");
      addClient(connectionManager, "client-3");

      connectionManager.setClientPurpose("client-1", "primary");
      connectionManager.setClientPurpose("client-2", "probe");
      connectionManager.setClientPurpose("client-3", "primary");

      const primaryClients = connectionManager.getPrimaryClients();
      expect(primaryClients).toHaveLength(2);
      expect(primaryClients.map((c) => c.id)).toEqual(
        expect.arrayContaining(["client-1", "client-3"]),
      );
    });

    it("should expose getProbeClients() returning only probe purpose clients", async () => {
      const connectionManager = new ConnectionManager();

      addClient(connectionManager, "client-1");
      addClient(connectionManager, "client-2");
      addClient(connectionManager, "client-3");

      connectionManager.setClientPurpose("client-1", "primary");
      connectionManager.setClientPurpose("client-2", "probe");
      connectionManager.setClientPurpose("client-3", "probe");

      const probeClients = connectionManager.getProbeClients();
      expect(probeClients).toHaveLength(2);
      expect(probeClients.map((c) => c.id)).toEqual(
        expect.arrayContaining(["client-2", "client-3"]),
      );
    });

    it("should include purpose in client metadata for diagnostics", async () => {
      const connectionManager = new ConnectionManager();

      addClient(connectionManager, "client-probe-1");
      connectionManager.setClientPurpose("client-probe-1", "probe");

      const client = connectionManager.getClient("client-probe-1");

      // MobileClient should have purpose field populated
      expect(client?.purpose).toBe("probe");
    });
  });

  describe("Connection purpose diagnostics logging", () => {
    it("should log purpose during auth for observability", async () => {
      const { handler, connectionManager } = createDependencies();
      const consoleSpy = jest.spyOn(console, "log").mockImplementation(() => {});

      await handler.handle(
        "client-diag-1",
        JSON.stringify({
          type: "auth",
          id: "auth-diag-1",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            client_version: "1.0.0",
            platform: "ios",
            purpose: "probe",
          },
        }),
      );

      // Bridge should log the purpose for diagnostics
      const loggedCalls = consoleSpy.mock.calls.map((call) => call.join(" "));
      const hasPurposeLog = loggedCalls.some(
        (log) => log.includes("client-diag-1") && log.includes("probe"),
      );

      expect(hasPurposeLog).toBe(true);

      consoleSpy.mockRestore();
    });
  });

  describe("Probe connection close code semantics", () => {
    it("should accept probe connections with documented close code expectation", async () => {
      // This test documents that probe connections are expected to close
      // with code 1000 (Normal) after health verification.
      // The actual close code validation happens at WebSocket layer,
      // this test verifies purpose metadata is correctly tagged.

      const { handler, sentMessages } = createDependencies();

      await handler.handle(
        "client-probe-close",
        JSON.stringify({
          type: "auth",
          id: "auth-probe-close",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            client_version: "1.0.0",
            platform: "ios",
            purpose: "probe",
          },
        }),
      );

      // Probe connection receives successful auth
      const ackMessage = sentMessages[0]?.message as BridgeMessage<ConnectionAckPayload>;
      expect(ackMessage.type).toBe("connection_ack");
      expect(ackMessage.payload.purpose).toBe("probe");

      // Note: Actual close code 1000 verification requires WebSocket mock
      // and is outside scope of this test slice.
    });
  });

  describe("Primary connection session association", () => {
    it("should allow session_start only from primary connections", async () => {
      const { handler, connectionManager, agentSdkAdapter } = createDependencies({
        existingPurpose: "probe",
      });

      // Authenticate first
      await handler.handle(
        "client-probe-session",
        JSON.stringify({
          type: "auth",
          id: "auth-probe-session",
          timestamp: new Date().toISOString(),
          payload: {
            token: "test-bridge-token",
            purpose: "probe",
          },
        }),
      );

      // Reset mocks to check session_start specifically
      jest.clearAllMocks();

      // Probe connection should not be allowed to start sessions
      await handler.handle(
        "client-probe-session",
        JSON.stringify({
          type: "session_start",
          id: "session-1",
          timestamp: new Date().toISOString(),
          payload: {
            agent: "claude-code",
            working_directory: "/repo",
          },
        }),
      );

      // session_start handler should NOT be called for probe connections
      // (or should return error indicating probe cannot start sessions)
      // This test expects the purpose field to be checked during session_start
      expect(agentSdkAdapter.handleSessionStart).not.toHaveBeenCalled();
    });
  });
});

File: apps/mobile/test/core/network/connection_purpose_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:recursor_mobile/core/network/websocket_messages.dart';
import 'package:recursor_mobile/core/network/websocket_service.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

/// Tests for connection purpose metadata handling (probe vs primary).
///
/// These tests define the expected behavior for mobile clients to declare
/// their connection purpose during auth handshake. Currently FAILING -
/// implementation code changes required.
///
/// Connection Purpose Semantics (per bridge-protocol.md):
/// - Probe: ~500ms-2s duration, health verification, capability check
/// - Primary: Session lifetime, active session for all communication
void main() {
  group('WebSocketService auth purpose field', () {
    test('sends purpose=probe in auth payload during health check flow',
        () async {
      final fakeChannel = FakeWebSocketChannel();
      final service = WebSocketService(channelFactory: (_) => fakeChannel);

      // Connect with purpose: probe for health verification
      final connectFuture = service.connect(
        url: 'wss://device.tailnet.ts.net:3000',
        token: 'bridge-token-123',
        purpose: ConnectionPurpose.probe, // NEW: Declare connection purpose
      );

      await Future<void>.delayed(Duration.zero);

      final authFrame =
          jsonDecode(fakeChannel.sentMessages.single) as Map<String, dynamic>;
      expect(authFrame['type'], 'auth');

      final payload = authFrame['payload'] as Map<String, dynamic>;
      expect(payload['token'], 'bridge-token-123');
      expect(payload['purpose'], 'probe'); // NEW: Purpose field in auth

      // Respond with connection_ack that echoes purpose
      fakeChannel.addIncoming(
        jsonEncode({
          'type': 'connection_ack',
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'payload': {
            'server_version': '1.0.0',
            'supported_agents': <String>[],
            'connection_mode': 'secure_remote',
            'connection_mode_description': 'Secure tunnel',
            'bridge_url': 'wss://device.tailnet.ts.net:3000',
            'requires_health_verification': true,
            'active_sessions': <Map<String, dynamic>>[],
            'purpose': 'probe', // Echo accepted purpose
          },
        }),
      );

      await connectFuture;

      // Service should track its declared purpose
      expect(service.connectionPurpose, ConnectionPurpose.probe);

      service.dispose();
      await fakeChannel.close();
    });

    test('sends purpose=primary in auth payload for session connections',
        () async {
      final fakeChannel = FakeWebSocketChannel();
      final service = WebSocketService(channelFactory: (_) => fakeChannel);

      final connectFuture = service.connect(
        url: 'wss://device.tailnet.ts.net:3000',
        token: 'bridge-token-456',
        purpose: ConnectionPurpose.primary, // Primary session connection
      );

      await Future<void>.delayed(Duration.zero);

      final authFrame =
          jsonDecode(fakeChannel.sentMessages.single) as Map<String, dynamic>;
      expect(authFrame['type'], 'auth');

      final payload = authFrame['payload'] as Map<String, dynamic>;
      expect(payload['purpose'], 'primary'); // NEW: Purpose field in auth

      fakeChannel.addIncoming(
        jsonEncode({
          'type': 'connection_ack',
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'payload': {
            'server_version': '1.0.0',
            'supported_agents': <String>[],
            'connection_mode': 'secure_remote',
            'connection_mode_description': 'Secure tunnel',
            'bridge_url': 'wss://device.tailnet.ts.net:3000',
            'requires_health_verification': true,
            'active_sessions': <Map<String, dynamic>>[],
            'purpose': 'primary',
          },
        }),
      );

      await connectFuture;

      expect(service.connectionPurpose, ConnectionPurpose.primary);

      service.dispose();
      await fakeChannel.close();
    });

    test('defaults to primary purpose when purpose argument omitted', () async {
      final fakeChannel = FakeWebSocketChannel();
      final service = WebSocketService(channelFactory: (_) => fakeChannel);

      // Legacy call without purpose argument - should default to primary
      final connectFuture = service.connect(
        url: 'wss://device.tailnet.ts.net:3000',
        token: 'bridge-token-789',
        // purpose argument omitted
      );

      await Future<void>.delayed(Duration.zero);

      final authFrame =
          jsonDecode(fakeChannel.sentMessages.single) as Map<String, dynamic>;
      final payload = authFrame['payload'] as Map<String, dynamic>;

      // Should default to 'primary' for backward compatibility
      expect(payload['purpose'], 'primary');

      fakeChannel.addIncoming(
        jsonEncode({
          'type': 'connection_ack',
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'payload': {
            'server_version': '1.0.0',
            'supported_agents': <String>[],
            'connection_mode': 'secure_remote',
            'connection_mode_description': 'Secure tunnel',
            'bridge_url': 'wss://device.tailnet.ts.net:3000',
            'requires_health_verification': true,
            'active_sessions': <Map<String, dynamic>>[],
            'purpose': 'primary',
          },
        }),
      );

      await connectFuture;

      expect(service.connectionPurpose, ConnectionPurpose.primary);

      service.dispose();
      await fakeChannel.close();
    });
  });

  group('BridgeMessage.auth factory includes purpose field', () {
    test('creates auth message with purpose from enum value', () async {
      final authMessage = BridgeMessage.auth(
        token: 'test-token',
        clientVersion: '1.0.0',
        platform: 'ios',
        purpose: ConnectionPurpose.probe, // NEW: purpose parameter
      );

      expect(authMessage.type, BridgeMessageType.auth);
      expect(authMessage.payload['token'], 'test-token');
      expect(authMessage.payload['client_version'], '1.0.0');
      expect(authMessage.payload['platform'], 'ios');
      expect(authMessage.payload['purpose'], 'probe'); // NEW field
    });

    test('creates auth message with primary purpose as default', () async {
      // Legacy call without purpose - should default to primary
      final authMessage = BridgeMessage.auth(
        token: 'test-token',
        clientVersion: '1.0.0',
        platform: 'android',
      );

      expect(authMessage.payload['purpose'], 'primary');
    });
  });

  group('Connection purpose state management', () {
    test('tracks active purpose in connection state', () async {
      final fakeChannel = FakeWebSocketChannel();
      final service = WebSocketService(channelFactory: (_) => fakeChannel);

      // Initially purpose is null before connection
      expect(service.connectionPurpose, isNull);

      final connectFuture = service.connect(
        url: 'wss://device.tailnet.ts.net:3000',
        token: 'bridge-token',
        purpose: ConnectionPurpose.primary,
      );

      await Future<void>.delayed(Duration.zero);

      // During connecting phase, purpose should be set
      expect(service.connectionPurpose, ConnectionPurpose.primary);

      fakeChannel.addIncoming(
        jsonEncode({
          'type': 'connection_ack',
          'timestamp': DateTime.now().toUtc().toIso8601String(),
          'payload': {
            'server_version': '1.0.0',
            'supported_agents': <String>[],
            'connection_mode': 'secure_remote',
            'connection_mode_description': 'Secure tunnel',
            'bridge_url': 'wss://device.tailnet.ts.net:3000',
            'requires_health_verification': true,
            'active_sessions': <Map<String, dynamic>>[],
          },
        }),
      );

      await connectFuture;
      service.dispose();
      await fakeChannel.close();
    });

    test('clears purpose on disconnect', () async {
      final fakeChannel = FakeWebSocketChannel();
      final service = WebSocketService(channelFactory: (_) => fakeChannel);

      final connectFuture = service.connect(
        url: 'wss://device.tailnet.ts.net:3000',
        token: 'bridge-token',
        purpose: ConnectionPurpose.primary,
      );

      await Future<void>.delayed(Duration.zero);

      fakeChannel.addIncoming(
        jsonEncode({
          'type': 'connection_ack',
          'payload': {
            'server_version': '1.0.0',
            'supported_agents': <String>[],
            'connection_mode': 'secure_remote',
            'connection_mode_description': 'Secure tunnel',
            'bridge_url': 'wss://device.tailnet.ts.net:3000',
            'requires_health_verification': true,
            'active_sessions': [],
            'purpose': 'primary',
          },
        }),
      );

      await connectFuture;

      expect(service.connectionPurpose, ConnectionPurpose.primary);

      service.disconnect();

      // Purpose should be cleared after disconnect
      expect(service.connectionPurpose, isNull);

      service.dispose();
      await fakeChannel.close();
    });
  });
}

/// Fake WebSocket channel for testing purposes.
///
/// Mirrors the implementation from websocket_service_test.dart for isolated testing.
class FakeWebSocketChannel implements WebSocketChannel {
  FakeWebSocketChannel({Future<void>? ready})
      : ready = ready ?? Future<void>.value(),
        _incomingController = StreamController<dynamic>.broadcast(),
        _sink = FakeWebSocketSink(<String>[]);

  final StreamController<dynamic> _incomingController;
  final FakeWebSocketSink _sink;

  @override
  final Future<void> ready;

  List<String> get sentMessages => _sink.messages;

  @override
  int? get closeCode => null;

  @override
  String? get closeReason => null;

  @override
  String? get protocol => null;

  @override
  Stream<dynamic> get stream => _incomingController.stream;

  @override
  WebSocketSink get sink => _sink;

  void addIncoming(String message) {
    _incomingController.add(message);
  }

  Future<void> close() async {
    await _sink.close();
    await _incomingController.close();
  }

  @override
  dynamic noSuchMethod(Invocation invocation) {
    return super.noSuchMethod(invocation);
  }
}

/// Fake WebSocket sink for testing purposes.
class FakeWebSocketSink implements WebSocketSink {
  FakeWebSocketSink(this.messages);

  final List<String> messages;
  final Completer<void> _doneCompleter = Completer<void>();

  @override
  Future<void> get done => _doneCompleter.future;

  @override
  void add(Object? data) {
    messages.add(data as String);
  }

  @override
  void addError(Object error, [StackTrace? stackTrace]) {}

  @override
  Future<void> addStream(Stream<dynamic> stream) async {
    await for (final data in stream) {
      add(data);
    }
  }

  @override
  Future<void> close([int? closeCode, String? closeReason]) async {
    if (!_doneCompleter.isCompleted) {
      _doneCompleter.complete();
    }
  }
}
